libraries/AP_HAL_Linux/Poller.cpp

```diff
@@ -0,0 +1,339 @@
+    }
+    if (ev & Poller::Event::Write) {
+        op |= EPOLLOUT;
+    }
+    if (ev & Poller::Event::Error) {
+        op |= EPOLLERR;
+    }
+
+    return op;
+}
+
+bool Poller::register_pollable(Pollable *p, const Poller::Event ev) {
+    if (_epfd < 0) {
+        return false;
+    }
+
+    struct epoll_event epev = {0};
+    epev.events = Poller::to_epoll_events(ev);
+    epev.data.ptr = static_cast<void *>(p);
+
+    return epoll_ctl(_epfd, EPOLL_CTL_ADD, p->get_fd(), &epev) == 0;
+}
+
+void Poller::unregister_pollable(const Pollable *p) {
+    if (_epfd >= 0) {
+        epoll_ctl(_epfd, EPOLL_CTL_DEL, p->get_fd(), NULL);
+    }
+}
+
+int Poller::poll(int timeout_ms) const {
+    const int max_events = 16;
+    epoll_event events[max_events];
+    const auto before_wait_us = AP_HAL::micros64();
+    const auto r = epoll_wait(_epfd, events, max_events, timeout_ms);
+    const auto delta_us = AP_HAL::micros64() - before_wait_us;
+    const auto delta_ms = delta_us / 1000;
+    const auto remaining_time_ms = timeout_ms - delta_ms;
+
+    if (r > 0) {
+        auto max_time_ms = remaining_time_ms / r;
+
+        if (!max_time_ms) {
+            max_time_ms = 1;
+        }
+
+        for (int i = 0; i < r; i++) {
+            Pollable *p = static_cast<Pollable *>(events[i].data.ptr);
+
+            if (events[i].events & EPOLLIN) {
+                p->on_can_read(max_time_ms);
+            }
+            if (events[i].events & EPOLLOUT) {
+                p->on_can_write(max_time_ms);
+            }
+            if (events[i].events & EPOLLERR) {
+                p->on_error(max_time_ms);
+            }
+            if (events[i].events & EPOLLHUP) {
+                p->on_hang_up(max_time_ms);
+            }
+        }
+    } else if (r < 0) {
+        if (errno == EINTR) {
+            // Try polling again with the remaining wait time.
+            return poll(remaining_time_ms);
+        }
+    }
+
+    return r;
+}
+
+Pollable::~Pollable() {
+    auto scheduler = Linux::Scheduler::from(hal.scheduler);
+    scheduler->unregister_pollable(this);
+
+    close(_fd);
+}
+
+bool Pollable::set_blocking(bool setting) {
+    auto curflags = fcntl(_fd, F_GETFL, 0);
+
+    if (curflags < 0) {
+        return false;
+    }
+
+    if (setting) {
+        curflags &= ~O_NONBLOCK;
+    } else {
+        curflags |= O_NONBLOCK;
+    }
+
+    return fcntl(_fd, F_SETFL, curflags) == 0;
+}
+
+void BufferedPollable::on_can_read(int max_time_ms) {
+    if (!_read_sem.take(max_time_ms)) {
+        return;
+    }
+
+    ByteBuffer::IoVec vec[2];
+    const auto n_vec = _read_buffer.reserve(vec, _read_buffer.space());
+    if (n_vec) {
+        struct iovec iovec[n_vec];
+        for (int i = 0; i < n_vec; i++) {
+            iovec[i].iov_base = static_cast<void *>(vec[i].data);
+            iovec[i].iov_len = static_cast<size_t>(vec[i].len);
+        }
+
+        (void) readv(_fd, iovec, n_vec);
+        auto r = readv(_fd, iovec, n_vec);
```

Comment:
this looks wrong and looks like it was due to a ignore-return tentative. We have a `UNUSED_RETURN` macro now, but I think we should actually return if there's an error.